src/wdio/services/ExpressDevServerService.js

```
@@ -0,0 +1,106 @@
+    return ExpressDevServerService.compile(webpackConfig).then((fs) => {
+      const app = express();
+
+      // Setup a catch all route, we can't use 'static' because we need to use a virtual file system
+      app.get('*', (req, res, next) => {
+        let filename = req.url;
+        // Setup a default index for the server.
+        if (filename === '/') {
+          filename = `/${index}`;
+        }
+
+        const filepath = `${webpackConfig.output.path}${filename}`;
+
+        if (fs.existsSync(filepath)) {
+          res.setHeader('content-type', mime.contentType(path.extname(filename)));
+          res.send(fs.readFileSync(filepath));
+        } else {
+          next();
+        }
+      });
+
+      app.use((req, res, next) => {
+        const err = new Error(`Not Found: ${req.originalUrl}`);
+        err.status = 404;
+        next(err);
+      });
+
+      // error handler
+      app.use((err, req, res) => {
+        // set locals, only providing error in development
+        res.locals.message = err.message;
+        res.locals.error = err;
+
+        // render the error page
+        res.status(err.status || 500);
+        res.render('error');
+      });
+
+      const server = app.listen(port);
+      // eslint-disable-next-line no-console
+      console.log('[ExpresDevService] Express server started');
+
+      return server;
+    });
+  }
+
+  static compile(webpackConfig) {
+    return new Promise((resolve, reject) => {
+      const compiler = webpack(webpackConfig);
+      // setup a virtual file system to write webpack files to.
+      compiler.outputFileSystem = new MemoryFS();
+      // eslint-disable-next-line no-console
+      console.log('[ExpresDevService] Webpack compilation started');
+      compiler.run((err, stats) => {
+        if (err || stats.hasErrors()) {
+          // eslint-disable-next-line no-console
+          console.log('[ExpresDevService] Webpack compiled unsuccessfully');
+          reject();
```

Should we return the errors for easier debugging?

---

I'd say yes, debugging with minimal to no error output is always painful.

---

the err or stats would be too much output. Any errors seen here would be more easily seen by running webpack explicitly (or through webpack dev server) and not through this service. The webpack config used here is not modified in any way.